File: Stage27/Python/Problem2206.py
import sys
from collections import deque

input = sys.stdin.readline

dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

def bfs(N, M, graph):
    visited = [[[False] * 2 for _ in range(M)] for _ in range(N)]

    queue = deque()
    queue.append((0, 0, 1, 0))
    visited[0][0][0] = True

    while queue:
        x, y, dist, wall = queue.popleft()

        if x == N - 1 and y == M - 1:
            return dist

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx < 0 or ny < 0 or nx >= N or ny >= M:
                continue

            if graph[nx][ny] == 0 and not visited[nx][ny][wall]:
                visited[nx][ny][wall] = True
                queue.append((nx, ny, dist + 1, wall))

            if graph[nx][ny] == 1 and wall == 0 and not visited[nx][ny][1]:
                visited[nx][ny][1] = True
                queue.append((nx, ny, dist + 1, 1))

    return -1

def main():
    N, M = map(int, input().split())

    graph = []
    for _ in range(N):
        graph.append(list(map(int, input().strip())))

    print(bfs(N, M, graph))

if __name__ == "__main__":
    main()

File: Stage27/Java/Problem2206.java
package Stage27.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Problem2206 {
    static int N, M;
    static int[][] map;
    static boolean[][][] visited;

    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};

    static class Node {
        int x, y, dist, wall;

        Node(int x, int y, int dist, int wall) {
            this.x = x;
            this.y = y;
            this.dist = dist;
            this.wall = wall;
        }
    }

    static int bfs() {
        Queue<Node> queue = new LinkedList<>();
        queue.offer(new Node(0, 0, 1, 0));
        visited[0][0][0] = true;

        while (!queue.isEmpty()) {
            Node now = queue.poll();

            int x = now.x;
            int y = now.y;
            int dist = now.dist;
            int wall = now.wall;

            if (x == N - 1 && y == M - 1) {
                return dist;
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dx[i];

                if (nx < 0 || ny < 0 || nx >= N || ny >= M) continue;

                if (map[nx][ny] == 0 && !visited[nx][ny][wall]) {
                    visited[nx][ny][wall] = true;
                    queue.offer(new Node(nx, ny, dist + 1, wall));
                }

                if (map[nx][ny] == 1 && wall == 0 && !visited[nx][ny][1]) {
                    visited[nx][ny][1] = true;
                    queue.offer(new Node(nx, ny, dist + 1, 1));
                }
            }
        }
        return -1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        map = new int[N][M];
        visited = new boolean[N][M][2];

        for (int i = 0; i < N; i++) {
            String line = br.readLine();
            for (int j = 0; j < M; j++) {
                map[i][j] = line.charAt(j) - '0';
            }
        }

        System.out.println(bfs());
        br.close();
    }
}
